lib/msprime.c

```diff
@@ -2429,6 +2529,127 @@ msp_run_coalescent(msp_t *self, double max_time, unsigned long max_events)
+{
+    int ret = 0;
+    uint32_t N, i, j, k, p;
+    size_t segment_mem_offset = 0;
+    population_t *pop;
+    segment_t *u[2], *x;
+    segment_list_t **parents = NULL;
+    segment_list_t *segment_mem = NULL;
+    segment_list_t *s;
+    avl_node_t *a, *node;
+    avl_tree_t Q[2];
+
+    for ( i = 0; i < 2; i++ ) {
+        avl_init_tree(&Q[i], cmp_segment_queue, NULL);
+    }
+
+    for (j = 0; j < self->num_populations; j++) {
+
+        pop = &self->populations[j];
+        N = (uint32_t) avl_count(&pop->ancestors);
+        /* N = (uint32_t) msp_get_population_size(self, pop); */
+
+        // Allocate memory for linked list of offspring per parent
+        parents = calloc(N, sizeof(segment_list_t *));
+        segment_mem = malloc(msp_get_num_ancestors(self) *
+                             sizeof(segment_list_t));
+        if (parents == NULL || segment_mem == NULL){
+            ret = MSP_ERR_NO_MEMORY;
+            goto out;
+        }
+        // Iterate through ancestors and draw parents
+        for ( a = pop->ancestors.head; a != NULL; a = a->next ) {
+            s = segment_mem + segment_mem_offset;
+            segment_mem_offset++;
+            p = (uint32_t) gsl_rng_uniform_int(self->rng, N);
+            s->next = parents[p];
+            s->node = a;
+            parents[p] = s;
+        }
+
+        // Iterate through offspring of parent k, adding to avl_tree
+        for ( k = 0; k < N; k++) {
+            for (s = parents[k]; s != NULL; s = s->next) {
+                node = s->node;
+                x = (segment_t *) node->item;
+                avl_unlink_node(&pop->ancestors, node);
+                msp_free_avl_node(self, node);
+                
+                // Recombine ancestor
+                ret = msp_recombine(self, x, &u[0], &u[1]);
+                if (ret != 0) {
+                    goto out;
+                }
+                // Add to AVLTree for each parental chromosome
+                for ( i = 0; i < 2; i++) {
+                    if (u[i] != NULL) {
+                        ret = msp_priority_queue_insert(self, &Q[i], u[i]);
+                        if (ret != 0) {
+                            goto out;
+                        }
+                    }
+                }
+            }
+            // Merge segments in each parental chromosome
+            for ( i = 0; i < 2; i ++ ) {
+                ret = msp_merge_ancestors(self, &Q[i], (population_id_t) j);
+                if (ret != 0) {
+                    goto out;
+                }
+            }
+        }
+        free(parents);
+        free(segment_mem);
+        segment_mem = NULL;
+        parents = NULL;
+    }
+out:
+    if (parents != NULL) {
+        free(parents);
+    }
+    if (segment_mem != NULL){
+        free(segment_mem);
+    }
+    return ret;
+}
+
+/* The main event loop for the Wright Fisher model. */
+static int WARN_UNUSED
+msp_run_wright_fisher(msp_t *self, double max_time, unsigned long max_events)
```

Comment:
Rename to ``msp_dtwf_run``. It really helps when methods are named consistently. We should try to make all methods that go along the dtwf code path start with the ``msp_dtwf`` prefix.